### PR TITLE
Full review jussi

### PR DESCRIPTION
**Pull request by jusahah**
I followed this https://stackoverflow.com/a/72044946 to create a PR with all the code. For review and to have a place to discuss.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+)
+
+func validateCurrencyCode(CurrencyCode string) bool {
```
Comment (jusahah) [inline main.go:58]: Too verbose... can do just:

```
func validateCurrencyCode(CurrencyCode string) bool {
  if condition {
    return true
  }
  return false
}
```

File: main.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+func getRate(CurrencyFrom string, CurrencyTo string) float64 {
+	httpposturl := fmt.Sprintf("https://query1.finance.yahoo.com/v7/finance/chart/%v%v%v", CurrencyFrom, CurrencyTo, "=x?corsDomain=finance.yahoo.com&range=1d&interval=1d")
+	// todo: get rid of bytes.NewBuffer yet set a request header
+	var jsonData = []byte(`{
+		"b": "l"
+	}`)
+
+	request, error := http.NewRequest("GET", httpposturl, bytes.NewBuffer(jsonData))
```
Comment (jusahah) [inline main.go]: What is `jsonData` for? 

Just `request, error := http.NewRequest("GET", httpposturl, nil)` I expect should be ok.

Comment (jusahah) [reply main.go]: `httpposturl` -> `httpGetUrl` as it is a GET request.

File: main.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+	}
+	defer response.Body.Close()
+	if response.StatusCode != 200 {
+		fmt.Println("something was wrong. status code was not 200\n check https://en.wikipedia.org/wiki/ISO_4217 for valid currency codes")
+		os.Exit(response.StatusCode)
+	}
+	responseData, _ := ioutil.ReadAll(response.Body)
+	// digs into the json and pulls out the value we want
+	// todo: use std library to pull this value using structs
+	value := gjson.GetBytes(responseData, "chart.result.0.meta.regularMarketPrice")
```
Comment (jusahah) [inline main.go:85]: Yeah like `todo` says, you don't need a separate library for this, though maybe it does not hurt either. The `encoding/json` is fine.

Then you can parse the JSON into a struct like:

```
type YahooConversionResponse struct {
  // fields the response has...
}

var response YahooConversionResponse
json.Unmarshal([]byte(responseData), &response)
```

Not tested but something like that.

File: main.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	regularMarketPrice, _ := strconv.ParseFloat(value.String(), 64)
+	return regularMarketPrice
+
+}
+
+func main() {
+	start := time.Now()
+	Qty, _ := 1.0, 0
```
Comment (jusahah) [inline main.go]: `Qty := 1.0`

File: main.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+		// store key-value pairs in redis
+		redisSet(CurrencyPair, regularMarketPriceString)
+		redisSet(CurrencyPairInverse, regularMarketPriceInverseString)
+
+		// print rates
+		printRates(regularMarketPrice, Qty, CurrencyFrom, CurrencyTo)
+		printTally(regularMarketPrice, float64(Qty), CurrencyTo)
+	} else {
+		resultFloat64, _ := strconv.ParseFloat(result, 64)
```
Comment (jusahah) [inline main.go:132]: Just stash the raw `float` value into Redis? So you don't have to convert back-and-forth and can simplify code.

```
regularMarketPrice, err := redisGet(CurrencyPair)

if err != nil {
  regularMarketPrice := getRate(CurrencyFrom, CurrencyTo)
  redisSet(CurrencyPair, regularMarketPrice)
  redisSet(CurrencyPairInverse, 1/regularMarketPrice)
}

printRates(regularMarketPrice, Qty, CurrencyFrom, CurrencyTo)
printTally(regularMarketPrice, float64(Qty), CurrencyTo)
```

Something like that... simplifying the control flow.

File: README.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
+Currency converter
+=======================
+Converts between ISO 4217 currency codes. Takes 2 or 3 arguments. The last arg is opitonal. 
+
+Usage: go run main.go usd eur 100
```
Comment (jusahah) [inline README.md]: This command does not work.

`go build && ./currency-converter usd czk 100` works for building+running the code.

File: main.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+func main() {
+	start := time.Now()
+	Qty, _ := 1.0, 0
+	lenArgs := len(os.Args)
+	if lenArgs == 4 {
+		Qty, _ = strconv.ParseFloat(os.Args[3], 64)
+	}
+
+	if lenArgs < 3 {
+		fmt.Println("USAGE: con [currency_code] [currency_code] int")
```
Comment (jusahah) [inline main.go]: Remove `con`

File: main.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+	if CurrencyCode == "is in list of currencies" {
+		result = true
+	} else {
+		result = false
+	}
+	return result
+}
+
+func getRate(CurrencyFrom string, CurrencyTo string) float64 {
+	httpposturl := fmt.Sprintf("https://query1.finance.yahoo.com/v7/finance/chart/%v%v%v", CurrencyFrom, CurrencyTo, "=x?corsDomain=finance.yahoo.com&range=1d&interval=1d")
```
Comment (jusahah) [inline main.go]: Lolled at `corsDomain=finance.yahoo.com`... the endpoint is not meant for API use I guess?

**Review by jusahah — COMMENTED** [review]
Yeah very cool, few comments for code.

This is a script-type of app that exits a second or two later after the launch and uses zero concurrency mechanisms (everything is in single thread here). 

So keeping all the business logic in `main.go` is fine.

**Comment by jusahah** [comment]
@jessegersensonchess Few comments here and there :man_cartwheeling: Just what caught my eye given my Golang experience (which is low) and general code structuring experience (which is a bit less low).

**Comment by jusahah** [comment]
Btw check out https://github.com/golang-standards/project-layout for helping structure larger projects.

File: main.go (hunk 3 above)
Comment (jessegersensonchess) [reply main.go:85]: the yahoo response-fields struct was in there in an earlier version but I had a syntax issue. Will try again.

File: main.go (hunk 2 above)
Comment (jessegersensonchess) [reply main.go]: nice find. 
will remove the protocol and method and use `url`!?

File: main.go (hunk 8 above)
Comment (jessegersensonchess) [reply main.go]: :) 
the corsDomain variable was pulled from another API... not sure if it was on stackoverflow or github. Looks like it's not needed, will remove it:
`curl "https://query1.finance.yahoo.com/v7/finance/chart/USDCZK=x?range=1d&interval=1d"`

File: main.go (hunk 1 above)
Comment (jessegersensonchess) [reply main.go:58]: !!

File: main.go (hunk 2 above)
Comment (jessegersensonchess) [reply main.go]: just tested it, you were right:
nil is ok

File: main.go (hunk 5 above)
Comment (jessegersensonchess) [reply main.go:132]: Ok, I'll try that in a future PR. 
fwiw, far as I know, everything stored in Redis is type 'string'.

**Comment by jessegersensonchess** [comment]
@jusahah that was great, thanks.
Pushed the changes directly to your `full_review_jussi` branch.

**Pull request merged**